add-on/src/lib/ipfs-proxy/access-control.js

```diff
@@ -0,0 +1,73 @@
+'use strict'
+
+const EventEmitter = require('events')
+const PQueue = require('p-queue')
+
+class AccessControl extends EventEmitter {
+  constructor (storage, key = 'ipfsProxyAcl') {
+    super()
+    this._storage = storage
+    this._key = key
+    this._onStorageChange = this._onStorageChange.bind(this)
+    storage.onChanged.addListener(this._onStorageChange)
+    this._writeQ = new PQueue({ concurrency: 1 })
+  }
+
+  async _onStorageChange (changes) {
+    const isAclChange = Object.keys(changes).some((key) => key === this._key)
+    if (!isAclChange) return
+    const acl = await this.getAcl()
+    this.emit('change', acl)
+  }
+
+  async getAccess (origin, permission) {
+    const acl = await this.getAcl()
+    if (acl[origin] == null || acl[origin][permission] == null) return null
+    return { origin, permission, allow: acl[origin][permission] }
+  }
+
+  async setAccess (origin, permission, allow) {
+    return this._writeQ.add(async () => {
+      const access = { origin, permission, allow }
+      const acl = await this.getAcl()
+
+      acl[origin] = acl[origin] || {}
+      acl[origin][permission] = allow
```

Comment:
Guard against the **_forbidden origins_** like `__proto__`?  👻

Reply:
It's less pretty, but might be safer to structure the acl with data that comes from the outside as values rather than keys...

Reply:
I think a `Map` instead of an plain old object would be safer right?